packages/llmRunner/src/llmRunner/squiggleGenerator.ts

```diff
@@ -0,0 +1,330 @@
+import { generateSummary } from "./generateSummary";
+import { LLMName } from "./LLMClient";
+import { CodeState, LLMStepTemplate } from "./LLMStep";
+import { TimestampedLogEntry } from "./Logger";
+import {
+  diffCompletionContentToCodeState,
+  diffToNewCode,
+  generationCompletionContentToCodeState,
+  squiggleCodeToCodeStateViaRunningAndFormatting,
+} from "./processSquiggleCode";
+import {
+  adjustToFeedbackPrompt,
+  editExistingSquiggleCodePrompt,
+  generateNewSquiggleCodePrompt,
+} from "./prompts";
+import { Workflow } from "./Workflow";
+
+function addStepByCodeState(
+  workflow: Workflow,
+  codeState: CodeState,
+  prompt: string
+) {
+  const step =
+    codeState.type === "success"
+      ? // both of these steps take prompt+codeState
+        executeAdjustToFeedbackStep
+      : fixCodeUntilItRunsStep;
+
+  workflow.addStep(step, {
+    prompt: { kind: "prompt", value: prompt },
+    codeState: { kind: "codeState", value: codeState },
+  });
+}
+
+export interface LlmConfig {
+  llmName: LLMName;
+  priceLimit: number;
+  durationLimitMinutes: number;
+  messagesInHistoryToKeep: number;
+}
+
+export const llmConfigDefault: LlmConfig = {
+  llmName: "Claude-Sonnet",
+  priceLimit: 0.3,
+  durationLimitMinutes: 1,
+  messagesInHistoryToKeep: 4,
+};
+
+export interface SquiggleResult {
+  isValid: boolean;
+  code: string;
+  logs: TimestampedLogEntry[];
+  totalPrice: number;
+  runTimeMs: number;
+  llmRunCount: number;
+  logSummary: string; // markdown
+}
+
+export type GeneratorInput =
+  | { type: "Create"; prompt: string }
+  | { type: "Edit"; code: string };
+
+const codeGeneratorStep = new LLMStepTemplate(
+  "GenerateCode",
+  {
+    inputs: { prompt: "prompt" },
+    outputs: { code: "code" },
+  },
+  async (context) => {
+    const prompt = context.getInput("prompt").value;
+    const promptPair = generateNewSquiggleCodePrompt(prompt);
+    const completion = await context.queryLLM(promptPair);
+
+    if (completion) {
+      const state = await generationCompletionContentToCodeState(completion);
+      if (state.okay) {
+        addStepByCodeState(context.workflow, state.value, prompt);
```

Comment:
Minor, but I think I prefer this being attached to this step instead. Like, "recommendedNextStep". Seems weird to mutate the workflow itself, as you mentioned.